activerecord/lib/active_record/connection_adapters/abstract/quoting.rb

```diff
@@ -157,6 +157,14 @@ def prepare_binds_for_database(binds) # :nodoc:
 
       private
 
+      def type_casted_binds(binds)
+        if binds.first.is_a?(Array)
+          binds.map { |column, value| type_cast(value, column) }
+        else
+          binds.map { |attr| type_cast(attr.value_for_database) }
```

Comment:
We have a method above: `prepare_binds_for_database` that does exactly what this does.